docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+Tendermint v0.35 or earlier to hard-fork when upgrading to v0.36.
+
+Since Solution 2 can be attacked, one might prefer Solution 3, even if it is more involved
+to implement. Further, we must elaborate on how we can turn Solution 3, described in abstract
+terms in the previous section, into a concrete implementation compatible with the current
+Tendermint codebase.
+
+### Current Limitations and Possible Implementations
+
+The main limitations affecting the current version of Tendermint are the following.
+
+- The current version of the blocksync reactor does not use the full
+  [light client verification](https://github.com/tendermint/tendermint/blob/4743a7ad0/spec/light-client/README.md) algorithm to validate blocks coming from other peers.
+- The code being structured into the blocksync and consensus reactors, only switching from the
+  blocksync reactor to the consensus reactor is supported; switching in the opposite direction is
+  not supported. Alternatively, the consensus reactor could have a mechanism allowing a late node
+  to catch up by skipping calls to `PrepareProposal`/`ProcessProposal`, and
+  `ExtendVote`/`VerifyVoteExtension` and only calling `FinalizeBlock` for each height.
+  Such a mechanism does not exist at the time of writing this RFC.
+
+The blocksync reactor featuring light client verification is being actively worked on (tentatively
+for v0.37). So it is best if this RFC does not try to solve that problem, but just makes sure
+the outcomes are compatible with that effort (TODO: check this paragraph with Callum/Jasmina).
+
+In subsection [Cases to Address](#cases-to-address), we concluded that we can focus on
+solving case (g) in theortical terms.
+However, as the current Tendermint version does not yet support switching back to blocksync once a
+node has switched to consensus, we need to split case (g) into two cases. When a full node needs to
+catch up...
+
+- **(g.1)** ... it has not switched yet from the blocksync reactor to the consensus reactor, or
+
+- **(g.2)** ... it has already switched to the consensus reactor.
+
+This is important in order to discuss the different possible implementations.
+
+#### Base Implementation: Persist and Propagate Extended Commit History
+
+In order to circumvent the fact that we cannot switch from the consensus reactor back to blocksync,
+rather than just keeping the few (TODO: 1?, 2?) most recent extended commits, nodes will need to keep
+and gossip a backlog of extended commits so that the consensus reactor can still propose and decide
+in out-of-date heights.
+
+The base implementation (for which an experimental patch exists) consists in the conservative
+approach of persisting in the block store *all* extended commits for which we have also stored
+the full block.
+Note that this does not apply to light blocks that might have been saved in the block store as a
+result of running statesync at startup.
```

Comment:
Would you mind expanding on this a bit? I think I'm not clear how light blocks apply here.

Reply:
My understanding of statesync is limited, but I saw in the code that statesync uses light block (i.e., the minimal information needed to light-client-verify the blockchain up to a certain height... I guess that of the snapshot). This minimal information, called signed header, contains the header and the commit, but e.g., not the transactions.

However, someone more knowledgeable on statesync might want to keep me honest here.

Anyway, back to your comment, I think I'll add a couple of sentences briefly explaining what a light block is and why we don't consider them.

Reply:
That's a good point. Let me add a TODO in the doc.
My first thought on this is: if a node statesyncs to _h_ in your example, that means:

* 2/3+ validators have got to _h+1_
* some node has made a snapshot of the state just after applying _h_

Now, independently of statesync, you need (at least) those 2/3+ validators to keep on making progress, otherwise the chain halts anyway.
So, with our "Base Implementation" (further below) we are good, as we will stay in blocksync until we have, at least, blocksynced one block.

Again, these are my first thoughts, so take it with a pinch of salt.

Reply:
@cason, after revisiting this, I see your comment is referring to the section describing the base solution. In this case there is no problem because we are protected by the exit condition imposed on blocksync (to switch to consensus). See [the code](https://github.com/tendermint/tendermint/blob/3377d616/internal/blocksync/reactor.go#L467), or the [text in this RFC](https://github.com/tendermint/tendermint/blob/8dc257af/docs/rfc/rfc-017-abci%2B%2B-vote-extension-propag.md?plain=1#L487) explaining it.

Also, in Solution 3 the condition that governs switching from consensus mode to catch-up mode, defined [here](https://github.com/tendermint/tendermint/blob/8dc257af/docs/rfc/rfc-017-abci%2B%2B-vote-extension-propag.md?plain=1#L398), is also protecting us against this case.

Let me know if you still have concerns.